sonar-text-plugin/src/main/resources/org/sonar/plugins/secrets/configuration/gcp/gcp.yaml

```diff
@@ -0,0 +1,65 @@
+        link: https://cwe.mitre.org/data/definitions/798
+        type: Standards
+      - description: MITRE, CWE-259 - Use of Hard-coded Password
+        link: https://cwe.mitre.org/data/definitions/259
+        type: Standards
+      - description: SANS Top 25 - Porous Defenses
+        link: https://www.sans.org/top25-software-errors/#cat3
+        type: Standards
+    message: Make sure this GCP secret gets revoked, changed, and removed from the code.
+  detection:
+    post:
+        patternNot: "(EXAMPLE|PRIVATE_KEY)"
+  rules:
+    - id: S6335
+      metadata:
+        name: Google Service Account Keys should not be disclosed
+      examples:
+        - text: >
+            {
+              "type": "service_account",
+              "project_id": "PROJECT_ID",
+              "private_key_id": "KEY_ID",
+              "private_key": "-----BEGIN PRIVATE KEY-----\\nPRIVATE_KEY \\n-----END PRIVATE KEY-----\\n",
+              "client_email": "SERVICE_ACCOUNT_EMAIL",
+              "client_id": "CLIENT_ID",
+              "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+              "token_uri": "https://accounts.google.com/o/oauth2/token",
+              "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
+              "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/SERVICE_ACCOUNT_EMAIL "
+            }
+          containsSecret: false
+        - text: >
+              {
+                "type": "service_account",
+                "project_id": "sonar-text-gcp-service-account-fake-project",
+                "private_key_id": "f560d248789641427bb72264839effb79098b86c",
+                "private_key": "-----BEGIN PRIVATE KEY-----\\nMIJEvQIBADANBgkqhkiG9w0BAQEFAASCBKcwggSjAgEAAoIBAQCoagGXuTHsB/r3\\n2W6raHa1Qrfj6pii5U2Ok53SxCyK3TxYc3Bfxq8orZeYC9LQ/I3tz7w4/BnT71AD\\nfP1i8SWHsRMIicSuVFcRoYMA+A1eNSmdrujdBNWgedfuSyHbPnNY7s8BBUIoBN7I\\n8gJG5DUUKAZfZDB2c/n7Yu0=\\n-----END PRIVATE KEY-----\\n",
+                "client_email": "gcp-service-account-key@fake-sonar-project.iam.gserviceaccount.com",
+                "client_id": "234567841521909194926",
+                "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+                "token_uri": "https://oauth2.googleapis.com/token",
+                "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
+                "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/gcp-service-account-key%40fake-sonar-project.iam.gserviceaccount.com",
+                "universe_domain": "googleapis.com"
+              }
+          containsSecret: true
+          match: "-----BEGIN PRIVATE KEY-----\\nMIJEvQIBADANBgkqhkiG9w0BAQEFAASCBKcwggSjAgEAAoIBAQCoagGXuTHsB/r3\\n2W6raHa1Qrfj6pii5U2Ok53SxCyK3TxYc3Bfxq8orZeYC9LQ/I3tz7w4/BnT71AD\\nfP1i8SWHsRMIicSuVFcRoYMA+A1eNSmdrujdBNWgedfuSyHbPnNY7s8BBUIoBN7I\\n8gJG5DUUKAZfZDB2c/n7Yu0=\\n-----END PRIVATE KEY-----"
```

Comment:
Ah if you try with regex101, it's because they flag it, even though it's a noncapturing group in the regex. Unfortunately, the existing tests that include it are way simpler and do not check what is matched, only whether something's raised

Reply:
You are right. I just tested and even if it should not match, it is in the final string:

``` bash
# output
-----BEGIN PRIVATE KEY-----\nMIJEvQIBADANBgkqhkiG9w0BAQEFAASCBKcwggSjAgEAAoIBAQCoagGXuTHsB/r3\n2W6raHa1Qrfj6pii5U2Ok53SxCyK3TxYc3Bfxq8orZeYC9LQ/I3tz7w4/BnT71AD\nfP1i8SWHsRMIicSuVFcRoYMA+A1eNSmdrujdBNWgedfuSyHbPnNY7s8BBUIoBN7I\n8gJG5DUUKAZfZDB2c/n7Yu0=\n-----END PRIVATE KEY-----\n
```

``` java

import java.util.ArrayList;
import java.util.List;
import java.util.regex.MatchResult;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.List;

public interface SecretsMatcher {
  List<Match> findIn(String content);
}

public class Match {
  private final String text;
  private final int fileStartOffset;
  private final int fileEndOffset;

  public Match(String text, int fileStartOffset, int fileEndOffset) {
    this.text = text;
    this.fileStartOffset = fileStartOffset;
    this.fileEndOffset = fileEndOffset;
  }

  public String getText() {
    return text;
  }

  public int getFileStartOffset() {
    return fileStartOffset;
  }

  public int getFileEndOffset() {
    return fileEndOffset;
  }
}

public class RegexMatcher implements SecretsMatcher {
  private final Pattern pattern;

  public RegexMatcher(String stringPattern) {
    this.pattern = Pattern.compile(stringPattern);
  }

  @Override
  public List<Match> findIn(String content) {
    List<Match> matches = new ArrayList<>();
    Matcher matcher = pattern.matcher(content);
    while (matcher.find()) {
      MatchResult matchResult = matcher.toMatchResult();
      matches.add(new Match(matchResult.group(1), matchResult.start(1), matchResult.end(1)));
    }
    return matches;
  }
}


public class Program
{
    public static void main(String[] args) {
		RegexMatcher regex = new RegexMatcher("\"private_key\"\\s*:\\s*\"(-----BEGIN PRIVATE KEY-----\\\\n[a-z-A-Z0-9+/=]{64}\\\\n" +
        "[a-z-A-Z0-9+/=\\\\]+-----END PRIVATE KEY-----(:?\\\\n)?)\""); 

        String content = """
                "type": "service_account",
                "project_id": "sonar-text-gcp-service-account-fake-project",
                "private_key_id": "f560d248789641427bb72264839effb79098b86c",
                "private_key": "-----BEGIN PRIVATE KEY-----\\nMIJEvQIBADANBgkqhkiG9w0BAQEFAASCBKcwggSjAgEAAoIBAQCoagGXuTHsB/r3\\n2W6raHa1Qrfj6pii5U2Ok53SxCyK3TxYc3Bfxq8orZeYC9LQ/I3tz7w4/BnT71AD\\nfP1i8SWHsRMIicSuVFcRoYMA+A1eNSmdrujdBNWgedfuSyHbPnNY7s8BBUIoBN7I\\n8gJG5DUUKAZfZDB2c/n7Yu0=\\n-----END PRIVATE KEY-----\\n",
                "client_email": "gcp-service-account-key@fake-sonar-project.iam.gserviceaccount.com",
                "client_id": "234567841521909194926",
                "auth_uri": "https://accounts.google.com/o/oauth2/auth",
                "token_uri": "https://oauth2.googleapis.com/token",
                "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
                "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/gcp-service-account-key%40fake-sonar-project.iam.gserviceaccount.com",
                "universe_domain": "googleapis.com"
            """;

        for (Match match : regex.findIn(content)) {
            System.out.println(match.getText());
        }
	}
}
```

Reply:
I added it